### ISSUE TITLE
[Bug] Node crash when applying a log entry to state machine

### ISSUE THREAD

**Issue by tdietert** [adjoint-io/raft#135]
# Problem

There is a bug in the subtle interplay between the validation of log entries by the leader and the application of new log entries to the ledger state. The condition that causes this issue is the fact that before replication, the leader validates commands against its current state machine state. Then, again during log application, all nodes in the network (leaders and followers alike) validate the command before applying it to their state machine. This causes a subtle issue when dealing with state machine command validation predicates that are order dependent; When commands are issued in quick succession, subsequent commands issued that are valid based on the current state machine can be invalid with respect to  previous log entries the leader has yet to fully replicate/commit. This can be understood with the following scenario:

1) Assume a key value store mapping variable names to integers as the state machine with the commands `insert`, `delete`, and `increment`, such that `increment "x"` and `delete "x"` are only valid commands when the variable `x` exists in the key value store.
2) Submit the command `insert "x" 7` to the state machine
3) The leader will validate this command before replicating it across all followers; When all followers apply this command to their state machine once replicated, this command is valid
4) Submit the commands `delete "x"` and `increment "x"` followed in quick succession such that the leader validates both with the state machine state `{ x = 7 }`.
5) When the leader replicates, commits, and applies both these log entries to its state machine, the application of the `increment "x"` command will fail during the validation step before applying the log to the state machine, as the variable "x" will have been removed from the state machine by the previous `delete` command.

I am currently experiencing this bug in an application from which I am using this raft library, and the leader node is crashing with: 
```
[2019-03-22T17:27:54.017+00:00] [2019-03-22T17:27:53:537](Debug) <127.0.0.1:8001 | Leader>: [Applying Log Entry]: Entry {entryIndex = Index 5, entryTerm = 1, entryValue = EntryValue (Transaction {header = TxAccount (CreateAccount {pubKey = "8509c538ae5b15f98257caeff7b1dd8d6591e6c38607b0095cea7e01d7492920a0c433ff304d5a91dba20407b2733bbebdd69c1fb2d6af1e7bf8f2aa7da9a660", timezone = "UTC", metadata = Metadata {unMetadata = fromList []}}), signature = Signature {sign_r = 87905625913392157655918524030828074465554514480282259884494452391509882606799, sign_s = 2503480248088869115538783095111134363675698102808764937630843331570769442806}, origin = Address "8r5eQr5w3gjfwuLQpDjpEGC2TEXSiS3yUZ7ShkW1UF79", tsData = Nothing}), entryIssuer = ClientIssuer (ClientId "127.0.0.1:8003:0:-511867297:42") (SerialNum 0), entryPrevHash = EntryHash {unEntryHash = "7681c10c40eeab439c1c31537c00968c86b4a177e03aca541670ab79e5c80fb9"}}
[2019-03-22T17:27:54.018+00:00] [2019-03-22T17:27:53:537](Critical) <127.0.0.1:8001 | Leader>: Failed to apply committed log entry: InvalidTransactions [InvalidTransaction {transaction = Transaction {header = TxAccount (CreateAccount {pubKey = "8509c538ae5b15f98257caeff7b1dd8d6591e6c38607b0095cea7e01d7492920a0c433ff304d5a91dba20407b2733bbebdd69c1fb2d6af1e7bf8f2aa7da9a660", timezone = "UTC", metadata = Metadata {unMetadata = fromList []}}), signature = Signature {sign_r = 87905625913392157655918524030828074465554514480282259884494452391509882606799, sign_s = 2503480248088869115538783095111134363675698102808764937630843331570769442806}, origin = Address "8r5eQr5w3gjfwuLQpDjpEGC2TEXSiS3yUZ7ShkW1UF79", tsData = Nothing}, reason = InvalidTransactionHeader (InvalidTransactionAccount (AccountError (AccountExists (Address "8r5eQr5w3gjfwuLQpDjpEGC2TEXSiS3yUZ7ShkW1UF79"))))}]
[2019-03-22T17:27:54.023+00:00] SupervisedChildRestarting {supervisorPid = pid://127.0.0.1:8001:0:26, childInScope = Just pid://127.0.0.1:8001:0:27, childSpecKey = "ConsensusRaftNode", exitReason = ExitOther "DiedException \"FatalError {fatalErrorMessage = \\\"logAndPanicIO: Failed to apply committed log entry: InvalidTransactions [InvalidTransaction {transaction = Transaction {header = TxAccount (CreateAccount {pubKey = \\\\\\\"8509c538ae5b15f98257caeff7b1dd8d6591e6c38607b0095cea7e01d7492920a0c433ff304d5a91dba20407b2733bbebdd69c1fb2d6af1e7bf8f2aa7da9a660\\\\\\\", timezone = \\\\\\\"UTC\\\\\\\", metadata = Metadata {unMetadata = fromList []}}), signature = Signature {sign_r = 87905625913392157655918524030828074465554514480282259884494452391509882606799, sign_s = 2503480248088869115538783095111134363675698102808764937630843331570769442806}, origin = Address \\\\\\\"8r5eQr5w3gjfwuLQpDjpEGC2TEXSiS3yUZ7ShkW1UF79\\\\\\\", tsData = Nothing}, reason = InvalidTransactionHeader (InvalidTransactionAccount (AccountError (AccountExists (Address \\\\\\\"8r5eQr5w3gjfwuLQpDjpEGC2TEXSiS3yUZ7ShkW1UF79\\\\\\\"))))}]\\\"}\""}
```

The exception is being thrown from https://github.com/adjoint-io/raft/blob/master/src/Raft.hs#L579 because `applyLogEntry` is failing.

# Solution

It is natural to add state machine command validation to the raft library; this is not an issue regarding whether or not we should be validating these state machine commands when the leader receives them or again before applying the log entry to the state machine. 

The issue arises from the fact that state machine command validation should be done _with respect to the future state machine state to which it will be applied_. It seems to me that the solution is to maintain _two_ state machine states, one that is the current state of the system, and the second which is only maintained by the leader, namely the _prospective_ state machine if all log entries that are not committed were to be so. This way, **new commands that arrive before preceding log entries are committed will be validated against the state machine state as if the uncommitted log entries we're already applied to the state machine**. Unfortunately, this will effectively double the memory needed to store the current state machine state, but it seems to me to be the only viable solution.

The only other obvious solution is to remove state machine command validation, which seems like the incorrect path forward; Many applications that will use this library have shared states that require some form of validation such that commands that are valid given a current state can be made invalid if preceding uncommitted entries are applied to the state machine.

**Comment by tdietert** [comment]
# Solution (?)

There is, potentially, another solution that we could explore: Allow invalid log entries to exist in the log. Upon application of log entries to the log, if they are not valid simply ignore the log entry and do not apply it to the state machine. It may be best to log the invalid log entry application failure here. This failure will not happen regularly, as the throughput of many application is not so high, and most state machine commands are _usually_ not conflicting with command submitted around the same time; e.g. (the example given in the initial text of this issue) two nodes are most likely not trying to delete/increment variables at the same time for the most part. 

Unfortunately, this is not necessarily a simpler solution that the previous one proposed, because of the way in which client write requests are responded to: In this implementation, the leader would send a "success" response to the client after the command was replicated on a majority of nodes, however, the command would actually have been invalid. For this solution to work, we would have to decide on a way to notify the client that although the command was originally accepted, it was invalid when applied to the state machine.

**Comment by tdietert** [comment]
Closing, solved by #136
